nistats/permutation_tests.py

```diff
@@ -0,0 +1,296 @@
+    if len(np.where(con_val != 0.0)[0]) != 2:
+        return False
+    return True
+
+
+def _permutations_glm(original_stat, glm_ref, imgs, design_matrix_objs,
+                      con_val, con_id, con_labels, n_perm_chunk=10000,
+                      two_sided_test=True,
+                      random_state=None,
+                      n_perm=10000,
+                      thread_id=1,
+                      verbose=0):
+    """Massively univariate group analysis with permuted OLS on a data chunk.
+    To be used in a parallel computing context.
+    Parameters
+    ----------
+    design_matrix_objs: list of [time_frames, paradigm, kwargs dict].
+        frame_times is an array of shape (n_frames,) representing the timing
+        of the scans in seconds. Paradigm is a DataFrame instance with the
+        description of the experimental paradigm. kwargs are any other
+        arguments that could be passed to the function make_design_matrix
+        defined in nistats.design_matrix.
+
+    original_stat : array-like, shape=(n_voxels)
+        t-scores obtained for the original (non-permuted) data.
+
+    n_perm_chunk : int,
+        Number of permutations to be performed.
+
+    two_sided_test : boolean,
+        If True, performs an unsigned t-test. Both positive and negative
+        effects are considered; the null hypothesis is that the effect is zero.
+        If False, only positive effects are considered as relevant. The null
+        hypothesis is that the effect is zero or negative.
+
+    random_state : int or None,
+        Seed for random number generator, to have the same permutations
+        in each computing units.
+
+    Returns
+    -------
+    unc_ranks_parts: array-like, shape=(n_voxels)
+        Accumulated count of permuted stats over original stats,
+        to compute uncorrected p-values. (limited to this permutation chunk).
+
+    smax_parts : array-like, shape=(n_perm_chunk, )
+        Distribution of the (max) t-statistic under the null hypothesis
+        (limited to this permutation chunk).
+
+    References
+    ----------
+    [1] Nichols, T. E., & Holmes, A. P. (2002). Nonparametric permutation
+    tests for functional neuroimaging: a primer with examples. Human brain
+    mapping, 15(1), 1-25..
+    """
+    # initialize the seed of the random generator
+    rng = check_random_state(random_state)
+
+    # run the permutations
+    smax_parts = np.empty((n_perm_chunk))
+    unc_rank_parts = np.zeros(len(original_stat))
+    # Infer conditions to permute
+    conditions = np.where(con_val != 0.0)
+    t0 = time.time()
+    for i in range(n_perm_chunk):
+        # Shuffle labels of the conditions in the design matrix
+        design_matrices = []
+
+        for frame_times, paradigm, kwargs in design_matrix_objs:
+            perm_paradigm = paradigm.copy()
+            con_idx = np.where(con_val != 0.0)[0]
+            conditions = [con_labels[x] for x in con_idx]
```

Comment:
Note that providing a list of identifiers rather than a contrast would actuallay make it possible to permute more than 2 items.